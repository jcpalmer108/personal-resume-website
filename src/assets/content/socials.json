{
  "links": [
    {
      "key": "github",
      "label": "Github",
      "url": "https://www.github.com/jcpalmer108/"
    },
    {
      "key": "leetcode",
      "label": "Leetcode",
      "url": "https://leetcode.com/darthToaster/"
    },
    {
      "key": "linkedin",
      "label": "LinkedIn",
      "url": "https://www.linkedin.com/in/jenna-palmer/"
    },
    {
      "key": "twitter",
      "label": "Twitter",
      "url": "https://twitter.com/jpalmerproject/"
    },
    {
      "key": "email",
      "label": "Email",
      "url": "mailto:jenna@jennapalmer.info"
    }
  ]
}
