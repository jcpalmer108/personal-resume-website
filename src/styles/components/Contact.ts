import styled from "styled-components"
import { redstone, silver_polish, tabletBreakpoint } from "../variables"
// import { silver_polish } from "../variables"

export const Mobile = styled.div`
  text-align: center;

  @media (min-width: ${tabletBreakpoint}px) {
    display: none;
  }
`

export const NotMobile = styled.div`
  display: none;

  @media (min-width: ${tabletBreakpoint}px) {
    display: flex;
    flex-direction: row;
    align-items: center;
  }
`

export const Form = styled.form`

`

export const InputWrapper = styled.form`

`


export const LeftWrapper = styled.div`
  padding: 35px 25px;
`


export const Title = styled.h2`

  @media (min-width: ${tabletBreakpoint}px) {
    width: 80%;
  }
`

export const Description = styled.p`

`



export const Left = styled.div`
  width: 55%;
  border: 1px solid ${silver_polish};
  `

export const Right = styled.div`
  flex-grow: 1;
  border: 1px solid ${silver_polish};
  border-left: none;
  height: 100%;

  `

export const Button = styled.input`
  border: none;
  background: ${redstone};
  width: 100%;
  font-size: 12px;
  color: white;
  height: 50px;
  font-family: Gilroy-Medium;
  font-size: 12px;
  line-spacing: 27px;
  letter-spacing: 2px;

  &:disabled {
    opacity: 0.3;
  }
`