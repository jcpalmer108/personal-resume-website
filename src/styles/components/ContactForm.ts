import styled from "styled-components";
import { silver_polish, tabletBreakpoint } from "../variables";

export const Wrapper = styled.div`
  border-bottom: 0px;

  @media (min-width: ${tabletBreakpoint}px) {
    border-bottom: 1px solid ${silver_polish};

    &:last-child {
      border-bottom: 0px;
    }
  }
`;
