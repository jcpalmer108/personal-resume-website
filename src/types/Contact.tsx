import { Sections, ContactInfo } from "./content"

export type ContactProps = {
  content: Sections | undefined,
  contact: ContactInfo,
  formFieldValues?: FormProps
}

export type FormProps = {
  name?: String,
  phone?: String,
  email?: String,
  message?: String
}
