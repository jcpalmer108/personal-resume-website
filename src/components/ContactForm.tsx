import { ContactFormProps } from "../types/ContactForm";
import Input from "./Input";

export default function ContactForm({
  formValues,
  updateForm,
}: ContactFormProps) {
  return (
    <div data-testid="ContactForm">
      <Input
        label="Name"
        value={formValues.name}
        updateForm={(e: any) => updateForm("name", e)}
      />
      <Input
        label="Phone"
        value={formValues.phone}
        updateForm={(e: any) => updateForm("phone", e)}
      />
      <Input
        label="Email"
        value={formValues.email}
        updateForm={(e: any) => updateForm("email", e)}
      />
      <Input
        area
        label="Message"
        value={formValues.message}
        updateForm={(e: any) => updateForm("message", e)}
      />
    </div>
  );
}
