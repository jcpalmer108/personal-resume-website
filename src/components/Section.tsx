import { 
  Wrapper, 
  Header, 
  Line, 
  Label, 
  Title, 
  Children
} from "../styles/components/Section";
import line from "../assets/images/label-line.svg";
import { SectionProps } from "../types/Section"

export default function Section({ 
  label,
  children,
  light,
  thin,
  wide,
  center,
  bottom,
  noTop,
  noBorder
}: SectionProps) {
  return (
    <Wrapper noBorder={noBorder} light={light} thin={thin} bottom={bottom} data-testid="Wrapper">
      {label && (
        <Header noTop={noTop} data-testid="Header" center={center} >
          <Label data-testid="Label">
            <Line src={line} alt="test"/>
            <Title>{label.toUpperCase()}</Title>
          </Label>
        </Header>
      )}
      <Children padded={!wide} data-testid="Children">{children}</Children>
    </Wrapper>
  )
}