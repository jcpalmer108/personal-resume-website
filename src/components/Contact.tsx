import { useState } from "react";
import socials from "../assets/content/socials.json";
import {
  BlankSection,
  BottomBar,
  Button,
  FormWrapper,
  Icon,
  InputWrapper,
  Left,
  LeftWrapper,
  Links,
  Mobile,
  NotMobile,
  Right,
  Socials,
  Title,
} from "../styles/components/Contact";
import { ContactProps, FormProps } from "../types/Contact";
import Section from "./Section";
import ContactForm from "./ContactForm";

export default function Contact({
  content,
  contact,
  formFieldValues,
}: ContactProps) {
  const [formValues, setFormValues] = useState<FormProps>(
    formFieldValues || { name: "", phone: "", email: "", message: "" }
  );
  const [isDisabled, setIsDisabled] = useState<boolean>(
    formValues.name === "" ||
      formValues.phone === "" ||
      formValues.email === "" ||
      formValues.message === ""
  );

  if (!content || !contact) {
    return null;
  }

  const updateFormValues = (key: string, e: any) => {
    setFormValues({
      ...formValues,
      [key]: e.target.value,
    });
    updateIsDisabled();
  };

  const generateBody = () => {
    const body = [
      `Phone: ${formValues.phone}`,
      `Email: ${formValues.email}\n`,
      `Message: ${formValues.message}`,
    ];

    return encodeURI(body.join("\n"));
  };

  const resetFormValues = () => {
    setFormValues({
      name: "",
      phone: "",
      email: "",
      message: "",
    });
  };

  const openMailLink = (e: any) => {
    e.preventDefault();
    const emailSubject = encodeURI("NEW: Website Response");
    const url = `mailto:${
      contact.email
    }?subject=${emailSubject}&body=${generateBody()}`;
    window.open(url, "_blank");
    resetFormValues();
  };

  const updateIsDisabled = () => {
    const { name, phone, email, message } = formValues;
    setIsDisabled(
      name === "" || phone === "" || email === "" || message === ""
    );
  };

  return (
    <div id="contact" data-testid="Contact">
      <Section wide>
        <Mobile data-testid="Mobile" onSubmit={openMailLink}>
          <Section label={content?.label} center>
            <Title>{content?.title}</Title>
            <p>{content?.description && content?.description[0]}</p>
            <ContactForm
              formValues={formValues}
              updateForm={updateFormValues}
            />
            <Button
              data-testid="Submit"
              disabled={isDisabled}
              type="submit"
              value="SUBMIT"
            />
            <Socials>
              {socials.links.map((item) => (
                <Links data-testid="Links" key={item.key} href={item.url}>
                  <Icon
                    src={require("../assets/images/socials/" +
                      item.key +
                      ".svg")}
                    alt={item.label}
                  />
                </Links>
              ))}
            </Socials>
          </Section>
        </Mobile>
        <NotMobile data-testid="NotMobile">
          <FormWrapper>
            <Left>
              <Section noBorder noTop wide label={content?.label}>
                <LeftWrapper>
                  <Title data-testid="Title">{content?.title}</Title>
                  <p data-testid="Description">
                    {content?.description && content?.description[0]}
                  </p>
                </LeftWrapper>
              </Section>
            </Left>
            <Right>
              <InputWrapper>
                <ContactForm
                  formValues={formValues}
                  updateForm={updateFormValues}
                />
              </InputWrapper>
            </Right>
          </FormWrapper>
          <BottomBar>
            <Socials data-testid="Socials">
              {socials.links.map((item) => (
                <Links key={item.key} href={item.url}>
                  <Icon
                    src={require("../assets/images/socials/" +
                      item.key +
                      ".svg")}
                    alt={item.label}
                  />
                </Links>
              ))}
            </Socials>
            <Button
              data-testid="Submit"
              disabled={isDisabled}
              type="submit"
              value="SUBMIT"
            />
          </BottomBar>
        </NotMobile>
        <BlankSection data-testid="BlankSection" />
      </Section>
    </div>
  );
}
