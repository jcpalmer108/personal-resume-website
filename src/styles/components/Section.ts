import styled from "styled-components";
import { desktopBreakpoint, contentWidth, silver_polish, tabletBreakpoint } from "../variables";
import { WrapperProps, HeaderProps, ChildrenProps } from "../../types/Section"

export const Wrapper = styled.div<WrapperProps>`
  max-width: ${contentWidth}px;
  margin: auto;
  border-left: solid;
  border-right: solid;
  border-bottom: solid;
  border-color: transparent;
  border-width: 0px;

  @media (min-width: ${contentWidth}px) {
    ${props => props.noBorder ? `
      border: none;
    ` : `
      border-bottom: ${props.bottom ? "soild" : "none" };
      border-color: ${props.light ? "white" : silver_polish};
      border-width: ${props.thin ? 0.2 : 1}px;    
    `}
  }
`

export const Header = styled.div<HeaderProps>`
  width: fit-content;
  margin: ${props => props.center ? "auto" : "0px"};
  padding-top: 50px; 
  padding-left: 16px;
  
  @media (min-width: ${tabletBreakpoint}px) {
    margin: 0px;
    padding-left: 25px;
    padding-top: ${props => props.noTop ? '25px' : '80px'};
  }

  @media (min-width: ${desktopBreakpoint}px) {
    padding-left: 60px;
    padding-top: 85px;
  }
`

export const Label = styled.div`
  display: flex;
  column-gap: 10px;
  font-size: 11px;
  font-family: Gilroy-Regular;

  @media (min-width: ${tabletBreakpoint}px) {
    transform: rotate(-90deg);
    padding-bottom: calc(100% - 15px);
    font-size: 14px;
  }

  @media (min-width: ${desktopBreakpoint}px) {
    padding-bottom: calc(100% - 17px);
    font-size: 16px;
  }

`

export const Children = styled.div<ChildrenProps>`
  padding: ${props => props.padded ? "20px 15px" : "0px"};

  @media (min-width: ${tabletBreakpoint}px) {
    padding: ${props => props.padded ? "40px 25px" : "0px"};
  }

  @media (min-width: ${desktopBreakpoint}px) {
    padding: ${props => props.padded ? "45px 60px 55px 60px" : "0px"};
  }
`


export const Title = styled.span`
  font-size: 12px;

  @media (min-width: ${tabletBreakpoint}px) {
    font-size: 14px;
  }

  @media (min-width: ${desktopBreakpoint}px) {
    font-size: 16px;
  }

`

export const Line = styled.img`
  width: 25px;
`

