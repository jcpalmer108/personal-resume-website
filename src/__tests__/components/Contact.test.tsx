import { fireEvent, render, screen } from "@testing-library/react";
import socials from "../../assets/content/socials.json";
import Contact from "../../components/Contact";
import { ContactFormProps } from "../../types/ContactForm";
import { SectionProps } from "../../types/Section";

// mock content
jest.mock(
  "../../assets/content/socials.json",
  () => ({
    links: [
      {
        key: "github",
        label: "Github",
        url: "https://www.github.com/jcpalmer108/",
      },
      {
        key: "leetcode",
        label: "Leetcode",
        url: "https://leetcode.com/darthToaster/",
      },
    ],
  }),
  { virtual: true }
);

// mock components & spies
const mockSection = jest.fn();
jest.mock(
  "../../components/Section",
  () =>
    ({
      label,
      children,
      light,
      thin,
      wide,
      center,
      bottom,
      noTop,
      noBorder,
    }: SectionProps) => {
      mockSection(label, light, thin, wide, center, bottom, noTop, noBorder);
      return <div>{children}</div>;
    }
);

const mockContactForm = jest.fn();
jest.mock(
  "../../components/ContactForm",
  () =>
    ({ formValues, updateForm }: ContactFormProps) => {
      mockContactForm(formValues, updateForm ? true : false);
      return <div />;
    }
);

const spyWindowOpen = jest.spyOn(window, "open");
spyWindowOpen.mockImplementation(jest.fn());

// params needed for component
const params = {
  required: {
    contact: {
      email: "jenna@jennapalmer.info",
      cell: "3142212451",
    },
  },
  optional: {
    content: {
      key: "contact",
      label: "test label",
      title: "test title",
      inMenu: true,
      description: ["test description"],
    },
    formValues: {
      name: "Jane Doe",
      phone: "1112223333",
      email: "test@test.com",
      message: "Hi there, I want to hire you!",
    },
  },
};

describe("Contact", () => {
  test("renders if content and contact are passed in", () => {
    // given
    const { required, optional } = params;
    render(<Contact content={optional.content} contact={required.contact} />);

    // then
    expect(screen.getByTestId("Contact")).toHaveAttribute("id", "contact");
    expect(screen.getByTestId("Mobile")).toMatchSnapshot();
    expect(screen.getByTestId("Mobile")).toHaveTextContent(
      optional.content.title
    );
    expect(screen.getByTestId("Mobile")).toHaveTextContent(
      optional.content.description[0]
    );
    expect(screen.getByTestId("NotMobile")).toMatchSnapshot();
    expect(screen.getByTestId("NotMobile")).toHaveTextContent(
      optional.content.title
    );
    expect(screen.getByTestId("NotMobile")).toHaveTextContent(
      optional.content.description[0]
    );
    expect(screen.getByTestId("BlankSection")).toMatchSnapshot();

    expect(mockContactForm).toHaveBeenCalledTimes(2);
    expect(mockContactForm).toHaveBeenNthCalledWith(
      1,
      {
        email: "",
        message: "",
        name: "",
        phone: "",
      },
      true
    );
    expect(mockContactForm).toHaveBeenNthCalledWith(
      2,
      {
        email: "",
        message: "",
        name: "",
        phone: "",
      },
      true
    );

    expect(screen.getAllByTestId("Submit")[0]).toHaveProperty("disabled", true);
    expect(screen.getByTestId("Socials").childNodes).toHaveLength(
      socials.links.length
    );
    socials.links.forEach((link, index) => {
      expect(screen.getByRole("link", { name: link.label })).toBeTruthy();
    });
  });

  test("does not render if only contact is passed in", () => {
    // given
    const { required } = params;
    render(<Contact content={undefined} contact={required.contact} />);

    // then
    expect(screen.queryByTestId("Contact")).toBeFalsy();
  });

  test("submit button is disabled unless all values are completed", () => {
    // given
    const { required, optional } = params;
    render(
      <Contact
        content={optional.content}
        contact={required.contact}
        formFieldValues={optional.formValues}
      />
    );

    // then
    expect(screen.getAllByTestId("Submit")[0]).toHaveProperty(
      "disabled",
      false
    );
  });

  test("submit button opens a mailto link with form data filled out", () => {
    // given
    const { required, optional } = params;
    render(
      <Contact
        content={optional.content}
        contact={required.contact}
        formFieldValues={optional.formValues}
      />
    );

    // when
    fireEvent.click(screen.getByRole("button", { name: "SUBMIT" }));

    // then
    expect(spyWindowOpen).toHaveBeenCalled();
    expect(spyWindowOpen).toHaveBeenCalledWith(
      "mailto:jenna@jennapalmer.info?subject=NEW:%20Website%20Response&body=Phone:%201112223333%0AEmail:%20test@test.com%0A%0AMessage:%20Hi%20there,%20I%20want%20to%20hire%20you!",
      "_blank"
    );
  });
});
