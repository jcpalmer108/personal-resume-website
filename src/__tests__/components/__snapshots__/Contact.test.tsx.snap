// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact renders if content and contact are passed in 1`] = `
<form
  class="sc-bczRLJ capNoS"
  data-testid="Mobile"
>
  <div>
    <h2
      class="sc-jSMfEi hlwoBT"
    >
      test title
    </h2>
    <p>
      test description
    </p>
    <div />
    <input
      class="sc-ftvSup byvgcr"
      data-testid="Submit"
      disabled=""
      type="submit"
      value="SUBMIT"
    />
    <div
      class="sc-jqUVSM jdMkJN"
    >
      <a
        class="sc-kDDrLX bMvUqc"
        data-testid="Links"
        href="https://www.github.com/jcpalmer108/"
      >
        <img
          alt="Github"
          class="sc-crXcEl lbtbww"
          src="[object Object]"
        />
      </a>
      <a
        class="sc-kDDrLX bMvUqc"
        data-testid="Links"
        href="https://leetcode.com/darthToaster/"
      >
        <img
          alt="Leetcode"
          class="sc-crXcEl lbtbww"
          src="[object Object]"
        />
      </a>
    </div>
  </div>
</form>
`;

exports[`Contact renders if content and contact are passed in 2`] = `
<form
  class="sc-gsnTZi kdzYlD"
  data-testid="NotMobile"
>
  <div
    class="sc-dkzDqf dGOnnx"
  >
    <div
      class="sc-gKXOVf cuiuTf"
    >
      <div>
        <div
          class="sc-eCYdqJ UUQIc"
        >
          <h2
            class="sc-jSMfEi hlwoBT"
            data-testid="Title"
          >
            test title
          </h2>
          <p
            data-testid="Description"
          >
            test description
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-iBkjds hGgFJv"
    >
      <div
        class="sc-hKMtZM jFrahT"
      >
        <div />
      </div>
    </div>
  </div>
  <div
    class="sc-papXJ qFTLl"
  >
    <div
      class="sc-jqUVSM jdMkJN"
      data-testid="Socials"
    >
      <a
        class="sc-kDDrLX bMvUqc"
        href="https://www.github.com/jcpalmer108/"
      >
        <img
          alt="Github"
          class="sc-crXcEl lbtbww"
          src="[object Object]"
        />
      </a>
      <a
        class="sc-kDDrLX bMvUqc"
        href="https://leetcode.com/darthToaster/"
      >
        <img
          alt="Leetcode"
          class="sc-crXcEl lbtbww"
          src="[object Object]"
        />
      </a>
    </div>
    <input
      class="sc-ftvSup byvgcr"
      data-testid="Submit"
      disabled=""
      type="submit"
      value="SUBMIT"
    />
  </div>
</form>
`;

exports[`Contact renders if content and contact are passed in 3`] = `
<div
  class="sc-iqcoie jyPsbp"
  data-testid="BlankSection"
/>
`;
